File: apps/operator_mobile/lib/src/features/scale/scale_adapter.dart
class ScaleSample {
  const ScaleSample({
    required this.grossKg,
    required this.tareKg,
    required this.stable,
  });

  final double grossKg;
  final double tareKg;
  final bool stable;

  double get netKg => grossKg - tareKg;
}

abstract class ScaleAdapter {
  Future<void> connect();
  Stream<ScaleSample> samples();
  Future<void> disconnect();
}

class SimulatedScaleAdapter implements ScaleAdapter {
  bool _connected = false;

  @override
  Future<void> connect() async {
    _connected = true;
  }

  @override
  Stream<ScaleSample> samples() async* {
    if (!_connected) throw StateError('Scale not connected');
    yield const ScaleSample(grossKg: 54.2, tareKg: 4.5, stable: false);
    await Future<void>.delayed(const Duration(milliseconds: 500));
    yield const ScaleSample(grossKg: 54.4, tareKg: 4.5, stable: false);
    await Future<void>.delayed(const Duration(milliseconds: 500));
    yield const ScaleSample(grossKg: 54.5, tareKg: 4.5, stable: true);
  }

  @override
  Future<void> disconnect() async {
    _connected = false;
  }
}

File: backend/app/weighment/scale_adapter.py
from abc import ABC, abstractmethod
from dataclasses import dataclass
from decimal import Decimal


@dataclass(frozen=True)
class ScaleSample:
    gross_kg: Decimal
    tare_kg: Decimal
    stable: bool

    @property
    def net_kg(self) -> Decimal:
        return self.gross_kg - self.tare_kg


class ScaleAdapter(ABC):
    """Vendor-neutral contract used by the Operator app/integration layer."""

    @abstractmethod
    def connect(self) -> None:
        raise NotImplementedError

    @abstractmethod
    def read_sample(self) -> ScaleSample:
        raise NotImplementedError

    @abstractmethod
    def disconnect(self) -> None:
        raise NotImplementedError


class SimulatedScaleAdapter(ScaleAdapter):
    def __init__(self, gross_kg: Decimal, tare_kg: Decimal, stable: bool = True):
        self.sample = ScaleSample(gross_kg=gross_kg, tare_kg=tare_kg, stable=stable)
        self.connected = False

    def connect(self) -> None:
        self.connected = True

    def read_sample(self) -> ScaleSample:
        if not self.connected:
            raise RuntimeError("scale not connected")
        return self.sample

    def disconnect(self) -> None:
        self.connected = False
